### ISSUE TITLE
SQL: Constraint failed

### ISSUE THREAD

**Issue by benbucksch** [mustang-im/mustang#209]
### Reproduction

* Unknown
* Might be a race condition
* It appears that the message got deleted while we're still trying to save the last parts of it.

### Actual result

![image](https://github.com/user-attachments/assets/6a024bf5-d8fe-4de3-a9af-d521bb6b4b3c)
![image](https://github.com/user-attachments/assets/c9503a15-3b30-4a1f-b8de-e6632637eef1)

**Comment by jermy-c** [comment]
**Reproduction**

1.  Click on an email folder
2. Delete an email (might be a step)
3. Click on another folder
4. Click get mail
5. Repeat 3-4

```
error.ts:6 Error: UNIQUE constraint failed: folder.accountID, folder.path
    at message.js:153:16
    at new Promise (<anonymous>)
    at WSCall.makeCall (message.js:148:12)
    at JPCWebSocket.callRemote (protocol.js:161:31)
    at RemoteClass.run (obj.js:225:55)
    at SQLFolder.save (SQLFolder.ts:27:48)
    at async SQLStorage.saveFolder (SQLStorage.ts:20:5)
    at async OWAFolder.save (Folder.ts:159:5)
    at async EMailCollection.readFolder (EMailCollection.ts:16:7)
    at async OWAFolder.readFolder (Folder.ts:56:5)
```
```json
{
    "id": 1388,
    "success": false,
    "message": "UNIQUE constraint failed: folder.accountID, folder.path",
    "code": "SQLITE_CONSTRAINT_UNIQUE",
    "backendStack": "SqliteError: UNIQUE constraint failed: folder.accountID, folder.path\n    at Database.run (file:///Users/jeremyc/Documents/GitHub/mustang/e2/out/main/index.js:114665:38)\n    at JPCWebSocket.funcListener (file:///Users/jeremyc/Documents/GitHub/mustang/e2/out/main/index.js:46185:28)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async WSCall._incomingMessage (file:///Users/jeremyc/Documents/GitHub/mustang/e2/out/main/index.js:45750:20)\n    at async WebSocket.<anonymous> (file:///Users/jeremyc/Documents/GitHub/mustang/e2/out/main/index.js:45844:9)",
    "originalCall": {
        "id": 1388,
        "path": "func",
        "arg": {
            "obj": "28451409173558927360",
            "name": "run",
            "args": [
                {
                    "plainObject": {
                        "sourceParts": [
                            "\n        INSERT INTO folder(\n          accountID, name, path,\n          parent, specialUse\n        ) VALUES(\n          ",
                            ", ",
                            ", ",
                            ",\n          ",
                            ", ",
                            "\n        )"
                        ],
                        "parameters": [
                            15,
                            "Drafts",
                            "AQMkADAwATNiZmYAZS04NGI2LWM3ZjUtMDACLTAwCgAuAAADXZkUNRZ4V0af0o2MYFY/bQEAVDovpX83QEqh1FcYw0tq9AAAAgEPAAAA",
                            null,
                            "drafts"
                        ]
                    }
                }
            ]
        }
    }
}
```

**Comment by NeilRashbrook** [comment]
`Error: UNIQUE constraint failed: folder.accountID, folder.path` was a race condition when it happened before in #204. In that case a folder was being saved in the background and then again from an attempt to get mail but in this case it could just be that clicking on folders quickly enough causes the second save attempt to start before the first has succeeded.

One approach might be to ignore (literally!) the problem at the SQL level, by using `INSERT OR IGNORE` to ensure a row exists, then updating it (in case that was the purpose of the save), then getting the dbID (if wanted). Unfortunately `INSERT OR REPLACE` does not do what we want, but there is apparently also an `ON CONFLICT DO UPDATE` clause which might allow either the insert or the given update to succeed. I don't know how these clauses interact with the `lastInsertRowId` though, and this doesn't solve the other issue of saving and deleting an email at the same time.

Another approach might be to serialise database operations using a promise to keep track of whether an operation is taking place (unfortunately you can't necessarily optimise operations away so you still have to queue them). You do however have to be very careful not to introduce race conditions in the queueing code, and this still won't help if you, say, delete a folder while trying to save messages in it.

**Comment by benbucksch** [comment]
This appears to have been fixed in git commit 74267993 and followups.
